summingbird-storm/src/main/scala/com/twitter/summingbird/storm/builder/Topology.scala

```diff
@@ -0,0 +1,156 @@
+package com.twitter.summingbird.storm.builder
+
+import com.twitter.algebird.Semigroup
+import com.twitter.algebird.util.summer.Incrementor
+import com.twitter.summingbird.online.executor.{ InputState, OperationContainer }
+import com.twitter.summingbird.online.option.{ MaxEmitPerExecute, SummerBuilder }
+import com.twitter.summingbird.option.JobId
+import com.twitter.summingbird.storm.StormMetric
+import com.twitter.summingbird.storm.option.{ AckOnEntry, AnchorTuples, MaxExecutePerSecond }
+import com.twitter.tormenta.spout.{ Spout => TormentaSpout }
+import org.apache.storm.generated.StormTopology
+import org.apache.storm.topology.{ IRichBolt, TopologyBuilder }
+import org.apache.storm.{ Config => BacktypeStormConfig }
+import org.apache.storm.tuple.Tuple
+import scala.collection.{ Map => CMap }
+
+// Mark as package private?
+case class Topology(
+  spouts: Map[Topology.SpoutId[_], Topology.Spout[_]],
+  bolts: Map[Topology.BoltId[_, _], Topology.Bolt[_, _]],
+  edges: List[Topology.Edge[_]]
+) {
+  def withSpout[O](id: String, spout: Topology.Spout[O]): (Topology.SpoutId[O], Topology) = {
+    val spoutId = Topology.SpoutId[O](id)
+    assert(!spouts.contains(spoutId))
+    (spoutId, Topology(spouts.updated(spoutId, spout), bolts, edges))
+  }
+
+  def withBolt[I, O](id: String, bolt: Topology.Bolt[I, O]): (Topology.BoltId[I, O], Topology) = {
+    val boltId = Topology.BoltId[I, O](id)
+    assert(!bolts.contains(boltId))
+    (boltId, Topology(spouts, bolts.updated(boltId, bolt), edges))
+  }
+
+  def withEdge[T](edge: Topology.Edge[T]): Topology = {
+    assert(contains(edge.source) && contains(edge.dest))
+    assert(edge.source != edge.dest)
+    assert(edges.forall { anotherEdge =>
+      !(anotherEdge.source == edge.source && anotherEdge.dest == edge.dest)
+    })
+
+    Topology(spouts, bolts, edges ++ List(edge))
+  }
+
+  def contains(id: Topology.ComponentId): Boolean = id match {
+    case spoutId: Topology.SpoutId[_] => spouts.contains(spoutId)
+    case boltId: Topology.BoltId[_, _] => bolts.contains(boltId)
+  }
+
+  def incomingEdges[T](id: Topology.ReceivingId[T]): List[Topology.Edge[T]] = {
```

Comment:
I'm not sure I understand the question. If you ask should we use `ClassTag`s then I think we should but it's not simple because we don't keep this information for now in `Producer`.